File: recovered-source/jadx/sources/me/ag2s/epublib/util/commons/io/XmlStreamReader.java
package me.ag2s.epublib.util.commons.io;

import io.legado.app.constant.RSSKeywords;
import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.text.MessageFormat;
import java.util.Locale;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import me.ag2s.epublib.epub.PackageDocumentBase;

/* JADX INFO: loaded from: reader-pro-classes-3.2.14.jar:me/ag2s/epublib/util/commons/io/XmlStreamReader.class */
public class XmlStreamReader extends Reader {
    private static final int BUFFER_SIZE = 8192;
    private static final String UTF_8 = "UTF-8";
    private static final String US_ASCII = "US-ASCII";
    private static final String UTF_16 = "UTF-16";
    private static final String UTF_32 = "UTF-32";
    private final Reader reader;
    private final String encoding;
    private final String defaultEncoding;
    private static final String RAW_EX_1 = "Invalid encoding, BOM [{0}] XML guess [{1}] XML prolog [{2}] encoding mismatch";
    private static final String RAW_EX_2 = "Invalid encoding, BOM [{0}] XML guess [{1}] XML prolog [{2}] unknown BOM";
    private static final String HTTP_EX_1 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], BOM must be NULL";
    private static final String HTTP_EX_2 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], encoding mismatch";
    private static final String HTTP_EX_3 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], Invalid MIME";
    private static final ByteOrderMark[] BOMS = {ByteOrderMark.UTF_8, ByteOrderMark.UTF_16BE, ByteOrderMark.UTF_16LE, ByteOrderMark.UTF_32BE, ByteOrderMark.UTF_32LE};
    private static final String UTF_16BE = "UTF-16BE";
    private static final String UTF_16LE = "UTF-16LE";
    private static final String UTF_32BE = "UTF-32BE";
    private static final String UTF_32LE = "UTF-32LE";
    private static final String EBCDIC = "CP1047";
    private static final ByteOrderMark[] XML_GUESS_BYTES = {new ByteOrderMark("UTF-8", 60, 63, 120, 109), new ByteOrderMark(UTF_16BE, 0, 60, 0, 63), new ByteOrderMark(UTF_16LE, 60, 0, 63, 0), new ByteOrderMark(UTF_32BE, 0, 0, 0, 60, 0, 0, 0, 63, 0, 0, 0, 120, 0, 0, 0, 109), new ByteOrderMark(UTF_32LE, 60, 0, 0, 0, 63, 0, 0, 0, 120, 0, 0, 0, 109, 0, 0, 0), new ByteOrderMark(EBCDIC, 76, 111, 167, 148)};
    private static final Pattern CHARSET_PATTERN = Pattern.compile("charset=[\"']?([.[^; \"']]*)[\"']?");
    public static final Pattern ENCODING_PATTERN = Pattern.compile("<\\?xml.*encoding[\\s]*=[\\s]*((?:\".[^\"]*\")|(?:'.[^']*'))", 8);

    public String getDefaultEncoding() {
        return this.defaultEncoding;
    }

    public XmlStreamReader(final File file) throws IOException {
        this(new FileInputStream((File) Objects.requireNonNull(file)));
    }

    public XmlStreamReader(final InputStream inputStream) throws IOException {
        this(inputStream, true);
    }

    public XmlStreamReader(final InputStream inputStream, final boolean lenient) throws IOException {
        this(inputStream, lenient, (String) null);
    }

    public XmlStreamReader(final InputStream inputStream, final boolean lenient, final String defaultEncoding) throws IOException {
        Objects.requireNonNull(inputStream, "inputStream");
        this.defaultEncoding = defaultEncoding;
        BOMInputStream bom = new BOMInputStream(new BufferedInputStream(inputStream, 8192), false, BOMS);
        BOMInputStream pis = new BOMInputStream(bom, true, XML_GUESS_BYTES);
        this.encoding = doRawStream(bom, pis, lenient);
        this.reader = new InputStreamReader(pis, this.encoding);
    }

    public XmlStreamReader(final URL url) throws IOException {
        this(((URL) Objects.requireNonNull(url, RSSKeywords.RSS_ITEM_URL)).openConnection(), (String) null);
    }

    public XmlStreamReader(final URLConnection conn, final String defaultEncoding) throws IOException {
        Objects.requireNonNull(conn, "conm");
        this.defaultEncoding = defaultEncoding;
        String contentType = conn.getContentType();
        InputStream inputStream = conn.getInputStream();
        BOMInputStream bom = new BOMInputStream(new BufferedInputStream(inputStream, 8192), false, BOMS);
        BOMInputStream pis = new BOMInputStream(bom, true, XML_GUESS_BYTES);
        if ((conn instanceof HttpURLConnection) || contentType != null) {
            this.encoding = processHttpStream(bom, pis, contentType, true);
        } else {
            this.encoding = doRawStream(bom, pis, true);
        }
        this.reader = new InputStreamReader(pis, this.encoding);
    }

    public XmlStreamReader(final InputStream inputStream, final String httpContentType) throws IOException {
        this(inputStream, httpContentType, true);
    }

    public XmlStreamReader(final InputStream inputStream, final String httpContentType, final boolean lenient, final String defaultEncoding) throws IOException {
        Objects.requireNonNull(inputStream, "inputStream");
        this.defaultEncoding = defaultEncoding;
        BOMInputStream bom = new BOMInputStream(new BufferedInputStream(inputStream, 8192), false, BOMS);
        BOMInputStream pis = new BOMInputStream(bom, true, XML_GUESS_BYTES);
        this.encoding = processHttpStream(bom, pis, httpContentType, lenient);
        this.reader = new InputStreamReader(pis, this.encoding);
    }

    public XmlStreamReader(final InputStream inputStream, final String httpContentType, final boolean lenient) throws IOException {
        this(inputStream, httpContentType, lenient, null);
    }

    public String getEncoding() {
        return this.encoding;
    }

    @Override // java.io.Reader
    public int read(final char[] buf, final int offset, final int len) throws IOException {
        return this.reader.read(buf, offset, len);
    }

    @Override // java.io.Reader, java.io.Closeable, java.lang.AutoCloseable
    public void close() throws IOException {
        this.reader.close();
    }

    private String doRawStream(final BOMInputStream bom, final BOMInputStream pis, final boolean lenient) throws IOException {
        String bomEnc = bom.getBOMCharsetName();
        String xmlGuessEnc = pis.getBOMCharsetName();
        String xmlEnc = getXmlProlog(pis, xmlGuessEnc);
        try {
            return calculateRawEncoding(bomEnc, xmlGuessEnc, xmlEnc);
        } catch (XmlStreamReaderException ex) {
            if (lenient) {
                return doLenientDetection(null, ex);
            }
            throw ex;
        }
    }

    private String processHttpStream(final BOMInputStream bom, final BOMInputStream pis, final String httpContentType, final boolean lenient) throws IOException {
        String bomEnc = bom.getBOMCharsetName();
        String xmlGuessEnc = pis.getBOMCharsetName();
        String xmlEnc = getXmlProlog(pis, xmlGuessEnc);
        try {
            return calculateHttpEncoding(httpContentType, bomEnc, xmlGuessEnc, xmlEnc, lenient);
        } catch (XmlStreamReaderException ex) {
            if (lenient) {
                return doLenientDetection(httpContentType, ex);
            }
            throw ex;
        }
    }

    private String doLenientDetection(String httpContentType, XmlStreamReaderException ex) throws IOException {
        if (httpContentType != null && httpContentType.startsWith("text/html")) {
            try {
                return calculateHttpEncoding("text/xml" + httpContentType.substring("text/html".length()), ex.getBomEncoding(), ex.getXmlGuessEncoding(), ex.getXmlEncoding(), true);
            } catch (XmlStreamReaderException ex2) {
                ex = ex2;
            }
        }
        String encoding = ex.getXmlEncoding();
        if (encoding == null) {
            encoding = ex.getContentTypeEncoding();
        }
        if (encoding == null) {
            encoding = this.defaultEncoding == null ? "UTF-8" : this.defaultEncoding;
        }
        return encoding;
    }

    String calculateRawEncoding(final String bomEnc, final String xmlGuessEnc, final String xmlEnc) throws IOException {
        if (bomEnc == null) {
            if (xmlGuessEnc == null || xmlEnc == null) {
                return this.defaultEncoding == null ? "UTF-8" : this.defaultEncoding;
            }
            if (xmlEnc.equals(UTF_16) && (xmlGuessEnc.equals(UTF_16BE) || xmlGuessEnc.equals(UTF_16LE))) {
                return xmlGuessEnc;
            }
            return xmlEnc;
        }
        if (bomEnc.equals("UTF-8")) {
            if (xmlGuessEnc != null && !xmlGuessEnc.equals("UTF-8")) {
                String msg = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg, bomEnc, xmlGuessEnc, xmlEnc);
            }
            if (xmlEnc != null && !xmlEnc.equals("UTF-8")) {
                String msg2 = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg2, bomEnc, xmlGuessEnc, xmlEnc);
            }
            return bomEnc;
        }
        if (bomEnc.equals(UTF_16BE) || bomEnc.equals(UTF_16LE)) {
            if (xmlGuessEnc != null && !xmlGuessEnc.equals(bomEnc)) {
                String msg3 = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg3, bomEnc, xmlGuessEnc, xmlEnc);
            }
            if (xmlEnc != null && !xmlEnc.equals(UTF_16) && !xmlEnc.equals(bomEnc)) {
                String msg4 = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg4, bomEnc, xmlGuessEnc, xmlEnc);
            }
            return bomEnc;
        }
        if (bomEnc.equals(UTF_32BE) || bomEnc.equals(UTF_32LE)) {
            if (xmlGuessEnc != null && !xmlGuessEnc.equals(bomEnc)) {
                String msg5 = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg5, bomEnc, xmlGuessEnc, xmlEnc);
            }
            if (xmlEnc != null && !xmlEnc.equals(UTF_32) && !xmlEnc.equals(bomEnc)) {
                String msg6 = MessageFormat.format(RAW_EX_1, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg6, bomEnc, xmlGuessEnc, xmlEnc);
            }
            return bomEnc;
        }
        String msg7 = MessageFormat.format(RAW_EX_2, bomEnc, xmlGuessEnc, xmlEnc);
        throw new XmlStreamReaderException(msg7, bomEnc, xmlGuessEnc, xmlEnc);
    }

    String calculateHttpEncoding(final String httpContentType, final String bomEnc, final String xmlGuessEnc, final String xmlEnc, final boolean lenient) throws IOException {
        if (lenient && xmlEnc != null) {
            return xmlEnc;
        }
        String cTMime = getContentTypeMime(httpContentType);
        String cTEnc = getContentTypeEncoding(httpContentType);
        boolean appXml = isAppXml(cTMime);
        boolean textXml = isTextXml(cTMime);
        if (!appXml && !textXml) {
            String msg = MessageFormat.format(HTTP_EX_3, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
            throw new XmlStreamReaderException(msg, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
        }
        if (cTEnc == null) {
            if (appXml) {
                return calculateRawEncoding(bomEnc, xmlGuessEnc, xmlEnc);
            }
            return this.defaultEncoding == null ? US_ASCII : this.defaultEncoding;
        }
        if (cTEnc.equals(UTF_16BE) || cTEnc.equals(UTF_16LE)) {
            if (bomEnc != null) {
                String msg2 = MessageFormat.format(HTTP_EX_1, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg2, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
            }
            return cTEnc;
        }
        if (cTEnc.equals(UTF_16)) {
            if (bomEnc != null && bomEnc.startsWith(UTF_16)) {
                return bomEnc;
            }
            String msg3 = MessageFormat.format(HTTP_EX_2, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
            throw new XmlStreamReaderException(msg3, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
        }
        if (cTEnc.equals(UTF_32BE) || cTEnc.equals(UTF_32LE)) {
            if (bomEnc != null) {
                String msg4 = MessageFormat.format(HTTP_EX_1, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
                throw new XmlStreamReaderException(msg4, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
            }
            return cTEnc;
        }
        if (cTEnc.equals(UTF_32)) {
            if (bomEnc != null && bomEnc.startsWith(UTF_32)) {
                return bomEnc;
            }
            String msg5 = MessageFormat.format(HTTP_EX_2, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
            throw new XmlStreamReaderException(msg5, cTMime, cTEnc, bomEnc, xmlGuessEnc, xmlEnc);
        }
        return cTEnc;
    }

    static String getContentTypeMime(final String httpContentType) {
        String mime;
        String mime2 = null;
        if (httpContentType != null) {
            int i = httpContentType.indexOf(";");
            if (i >= 0) {
                mime = httpContentType.substring(0, i);
            } else {
                mime = httpContentType;
            }
            mime2 = mime.trim();
        }
        return mime2;
    }

    static String getContentTypeEncoding(final String httpContentType) {
        int i;
        String encoding = null;
        if (httpContentType != null && (i = httpContentType.indexOf(";")) > -1) {
            String postMime = httpContentType.substring(i + 1);
            Matcher m = CHARSET_PATTERN.matcher(postMime);
            String encoding2 = m.find() ? m.group(1) : null;
            encoding = encoding2 != null ? encoding2.toUpperCase(Locale.ROOT) : null;
        }
        return encoding;
    }

    private static String getXmlProlog(final InputStream inputStream, final String guessedEnc) throws IOException {
        String encoding = null;
        if (guessedEnc != null) {
            byte[] bytes = new byte[8192];
            inputStream.mark(8192);
            int offset = 0;
            int max = 8192;
            int c = inputStream.read(bytes, 0, 8192);
            int firstGT = -1;
            String xmlProlog = PackageDocumentBase.PREFIX_OPF;
            while (c != -1 && firstGT == -1 && offset < 8192) {
                offset += c;
                max -= c;
                c = inputStream.read(bytes, offset, max);
                xmlProlog = new String(bytes, 0, offset, guessedEnc);
                firstGT = xmlProlog.indexOf(62);
            }
            if (firstGT == -1) {
                if (c == -1) {
                    throw new IOException("Unexpected end of XML stream");
                }
                throw new IOException("XML prolog or ROOT element not found on first " + offset + " bytes");
            }
            int bytesRead = offset;
            if (bytesRead > 0) {
                inputStream.reset();
                BufferedReader bReader = new BufferedReader(new StringReader(xmlProlog.substring(0, firstGT + 1)));
                StringBuffer prolog = new StringBuffer();
                while (true) {
                    String line = bReader.readLine();
                    if (line == null) {
                        break;
                    }
                    prolog.append(line);
                }
                Matcher m = ENCODING_PATTERN.matcher(prolog);
                if (m.find()) {
                    String encoding2 = ((String) Objects.requireNonNull(m.group(1))).toUpperCase(Locale.ROOT);
                    encoding = encoding2.substring(1, encoding2.length() - 1);
                }
            }
        }
        return encoding;
    }

    static boolean isAppXml(final String mime) {
        return mime != null && (mime.equals("application/xml") || mime.equals("application/xml-dtd") || mime.equals("application/xml-external-parsed-entity") || (mime.startsWith("application/") && mime.endsWith("+xml")));
    }

    static boolean isTextXml(final String mime) {
        return mime != null && (mime.equals("text/xml") || mime.equals("text/xml-external-parsed-entity") || (mime.startsWith("text/") && mime.endsWith("+xml")));
    }
}

File: src/main/kotlin/me/ag2s/epublib/util/commons/io/XmlStreamReader.kt
package me.ag2s.epublib.util.commons.io

import java.io.BufferedInputStream
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.Reader
import java.io.StringReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLConnection
import java.text.MessageFormat
import java.util.Locale
import java.util.Objects
import java.util.regex.Pattern

class XmlStreamReader : Reader {
    private val reader: Reader
    private val encoding: String
    private val defaultEncoding: String?
    fun getDefaultEncoding(): String? = defaultEncoding
    @Throws(IOException::class) constructor(file: File) : this(FileInputStream(Objects.requireNonNull(file)))
    @Throws(IOException::class) constructor(inputStream: InputStream) : this(inputStream, true)
    @Throws(IOException::class) constructor(inputStream: InputStream, lenient: Boolean) : this(inputStream, lenient, null)
    @Throws(IOException::class) constructor(inputStream: InputStream, lenient: Boolean, defaultEncoding: String?) {
        Objects.requireNonNull(inputStream, "inputStream")
        this.defaultEncoding = defaultEncoding
        val bom = BOMInputStream(BufferedInputStream(inputStream, BUFFER_SIZE), false, *BOMS)
        val pis = BOMInputStream(bom, true, *XML_GUESS_BYTES)
        encoding = doRawStream(bom, pis, lenient)
        reader = InputStreamReader(pis, encoding)
    }
    @Throws(IOException::class) constructor(url: URL) : this(Objects.requireNonNull(url, "url").openConnection(), null)
    @Throws(IOException::class) constructor(connection: URLConnection, defaultEncoding: String?) {
        Objects.requireNonNull(connection, "conm")
        this.defaultEncoding = defaultEncoding
        val contentType = connection.contentType
        val bom = BOMInputStream(BufferedInputStream(connection.inputStream, BUFFER_SIZE), false, *BOMS)
        val pis = BOMInputStream(bom, true, *XML_GUESS_BYTES)
        encoding = if (connection !is HttpURLConnection && contentType == null) doRawStream(bom, pis, true) else processHttpStream(bom, pis, contentType, true)
        reader = InputStreamReader(pis, encoding)
    }
    @Throws(IOException::class) constructor(inputStream: InputStream, httpContentType: String?) : this(inputStream, httpContentType, true)
    @Throws(IOException::class) constructor(inputStream: InputStream, httpContentType: String?, lenient: Boolean) : this(inputStream, httpContentType, lenient, null)
    @Throws(IOException::class) constructor(inputStream: InputStream, httpContentType: String?, lenient: Boolean, defaultEncoding: String?) {
        Objects.requireNonNull(inputStream, "inputStream")
        this.defaultEncoding = defaultEncoding
        val bom = BOMInputStream(BufferedInputStream(inputStream, BUFFER_SIZE), false, *BOMS)
        val pis = BOMInputStream(bom, true, *XML_GUESS_BYTES)
        encoding = processHttpStream(bom, pis, httpContentType, lenient)
        reader = InputStreamReader(pis, encoding)
    }
    fun getEncoding(): String = encoding
    @Throws(IOException::class) override fun read(buffer: CharArray, offset: Int, length: Int): Int = reader.read(buffer, offset, length)
    @Throws(IOException::class) override fun close() { reader.close() }

    @Throws(IOException::class) private fun doRawStream(bom: BOMInputStream, pis: BOMInputStream, lenient: Boolean): String {
        val bomEncoding = bom.getBOMCharsetName(); val guessEncoding = pis.getBOMCharsetName(); val xmlEncoding = getXmlProlog(pis, guessEncoding)
        return try { calculateRawEncoding(bomEncoding, guessEncoding, xmlEncoding) } catch (exception: XmlStreamReaderException) { if (lenient) doLenientDetection(null, exception) else throw exception }
    }
    @Throws(IOException::class) private fun processHttpStream(bom: BOMInputStream, pis: BOMInputStream, httpContentType: String?, lenient: Boolean): String {
        val bomEncoding = bom.getBOMCharsetName(); val guessEncoding = pis.getBOMCharsetName(); val xmlEncoding = getXmlProlog(pis, guessEncoding)
        return try { calculateHttpEncoding(httpContentType, bomEncoding, guessEncoding, xmlEncoding, lenient) } catch (exception: XmlStreamReaderException) { if (lenient) doLenientDetection(httpContentType, exception) else throw exception }
    }
    @Throws(IOException::class) private fun doLenientDetection(httpContentType: String?, initial: XmlStreamReaderException): String {
        var contentType = httpContentType; var exception = initial
        if (contentType != null && contentType.startsWith("text/html")) {
            contentType = "text/xml" + contentType.substring("text/html".length)
            try { return calculateHttpEncoding(contentType, exception.getBomEncoding(), exception.getXmlGuessEncoding(), exception.getXmlEncoding(), true) } catch (other: XmlStreamReaderException) { exception = other }
        }
        return exception.getXmlEncoding() ?: exception.getContentTypeEncoding() ?: defaultEncoding ?: UTF_8
    }
    @Throws(IOException::class) internal fun calculateRawEncoding(bomEncoding: String?, xmlGuessEncoding: String?, xmlEncoding: String?): String {
        if (bomEncoding != null) {
            if (bomEncoding == UTF_8) {
                if (xmlGuessEncoding != null && xmlGuessEncoding != UTF_8 || xmlEncoding != null && xmlEncoding != UTF_8) throw rawException(RAW_EX_1, bomEncoding, xmlGuessEncoding, xmlEncoding)
                return bomEncoding
            }
            if (bomEncoding == UTF_16BE || bomEncoding == UTF_16LE) {
                if (xmlGuessEncoding != null && xmlGuessEncoding != bomEncoding || xmlEncoding != null && xmlEncoding != UTF_16 && xmlEncoding != bomEncoding) throw rawException(RAW_EX_1, bomEncoding, xmlGuessEncoding, xmlEncoding)
                return bomEncoding
            }
            if (bomEncoding == UTF_32BE || bomEncoding == UTF_32LE) {
                if (xmlGuessEncoding != null && xmlGuessEncoding != bomEncoding || xmlEncoding != null && xmlEncoding != UTF_32 && xmlEncoding != bomEncoding) throw rawException(RAW_EX_1, bomEncoding, xmlGuessEncoding, xmlEncoding)
                return bomEncoding
            }
            throw rawException(RAW_EX_2, bomEncoding, xmlGuessEncoding, xmlEncoding)
        }
        return if (xmlGuessEncoding != null && xmlEncoding != null) if (xmlEncoding == UTF_16 && (xmlGuessEncoding == UTF_16BE || xmlGuessEncoding == UTF_16LE)) xmlGuessEncoding else xmlEncoding else defaultEncoding ?: UTF_8
    }
    @Throws(IOException::class) internal fun calculateHttpEncoding(httpContentType: String?, bomEncoding: String?, xmlGuessEncoding: String?, xmlEncoding: String?, lenient: Boolean): String {
        if (lenient && xmlEncoding != null) return xmlEncoding
        val mime = getContentTypeMime(httpContentType); val contentEncoding = getContentTypeEncoding(httpContentType); val appXml = isAppXml(mime); val textXml = isTextXml(mime)
        if (!appXml && !textXml) throw httpException(HTTP_EX_3, mime, contentEncoding, bomEncoding, xmlGuessEncoding, xmlEncoding)
        if (contentEncoding == null) return if (appXml) calculateRawEncoding(bomEncoding, xmlGuessEncoding, xmlEncoding) else defaultEncoding ?: US_ASCII
        if (contentEncoding == UTF_16BE || contentEncoding == UTF_16LE || contentEncoding == UTF_32BE || contentEncoding == UTF_32LE) { if (bomEncoding != null) throw httpException(HTTP_EX_1, mime, contentEncoding, bomEncoding, xmlGuessEncoding, xmlEncoding); return contentEncoding }
        if (contentEncoding == UTF_16) { if (bomEncoding != null && bomEncoding.startsWith(UTF_16)) return bomEncoding; throw httpException(HTTP_EX_2, mime, contentEncoding, bomEncoding, xmlGuessEncoding, xmlEncoding) }
        if (contentEncoding == UTF_32) { if (bomEncoding != null && bomEncoding.startsWith(UTF_32)) return bomEncoding; throw httpException(HTTP_EX_2, mime, contentEncoding, bomEncoding, xmlGuessEncoding, xmlEncoding) }
        return contentEncoding
    }
    private fun rawException(template: String, bom: String?, guess: String?, xml: String?): XmlStreamReaderException = XmlStreamReaderException(MessageFormat.format(template, bom, guess, xml), bom, guess, xml)
    private fun httpException(template: String, mime: String?, content: String?, bom: String?, guess: String?, xml: String?): XmlStreamReaderException = XmlStreamReaderException(MessageFormat.format(template, mime, content, bom, guess, xml), mime, content, bom, guess, xml)

    companion object {
        private const val BUFFER_SIZE = 8192
        private const val UTF_8 = "UTF-8"; private const val US_ASCII = "US-ASCII"; private const val UTF_16BE = "UTF-16BE"; private const val UTF_16LE = "UTF-16LE"; private const val UTF_32BE = "UTF-32BE"; private const val UTF_32LE = "UTF-32LE"; private const val UTF_16 = "UTF-16"; private const val UTF_32 = "UTF-32"
        private val BOMS = arrayOf(ByteOrderMark.UTF_8, ByteOrderMark.UTF_16BE, ByteOrderMark.UTF_16LE, ByteOrderMark.UTF_32BE, ByteOrderMark.UTF_32LE)
        private val XML_GUESS_BYTES = arrayOf(ByteOrderMark("UTF-8", 60, 63, 120, 109), ByteOrderMark(UTF_16BE, 0, 60, 0, 63), ByteOrderMark(UTF_16LE, 60, 0, 63, 0), ByteOrderMark(UTF_32BE, 0, 0, 0, 60, 0, 0, 0, 63, 0, 0, 0, 120, 0, 0, 0, 109), ByteOrderMark(UTF_32LE, 60, 0, 0, 0, 63, 0, 0, 0, 120, 0, 0, 0, 109, 0, 0, 0), ByteOrderMark("CP1047", 76, 111, 167, 148))
        private val charsetPattern = Pattern.compile("charset=[\"']?([.[^; \"']]*)[\"']?")
        @JvmField val ENCODING_PATTERN: Pattern = Pattern.compile("<\\?xml.*encoding[\\s]*=[\\s]*((?:\".[^\"]*\")|(?:'.[^']*'))", Pattern.CASE_INSENSITIVE)
        private const val RAW_EX_1 = "Invalid encoding, BOM [{0}] XML guess [{1}] XML prolog [{2}] encoding mismatch"
        private const val RAW_EX_2 = "Invalid encoding, BOM [{0}] XML guess [{1}] XML prolog [{2}] unknown BOM"
        private const val HTTP_EX_1 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], BOM must be NULL"
        private const val HTTP_EX_2 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], encoding mismatch"
        private const val HTTP_EX_3 = "Invalid encoding, CT-MIME [{0}] CT-Enc [{1}] BOM [{2}] XML guess [{3}] XML prolog [{4}], Invalid MIME"
        @JvmStatic internal fun getContentTypeMime(contentType: String?): String? = contentType?.let { it.substringBefore(';').trim() }
        @JvmStatic internal fun getContentTypeEncoding(contentType: String?): String? { val index = contentType?.indexOf(';') ?: -1; return if (index > -1) charsetPattern.matcher(contentType!!.substring(index + 1)).let { if (it.find()) it.group(1)?.uppercase(Locale.ROOT) else null } else null }
        @Throws(IOException::class) private fun getXmlProlog(input: InputStream, guessedEncoding: String?): String? {
            if (guessedEncoding == null) return null
            val bytes = ByteArray(BUFFER_SIZE); input.mark(BUFFER_SIZE); var offset = 0; var maximum = BUFFER_SIZE; var count = input.read(bytes, offset, maximum); var firstGt = -1; var prolog = ""
            while (count != -1 && firstGt == -1 && offset < BUFFER_SIZE) { offset += count; maximum -= count; count = input.read(bytes, offset, maximum); prolog = String(bytes, 0, offset, charset(guessedEncoding)); firstGt = prolog.indexOf('>') }
            if (firstGt == -1) { if (count == -1) throw IOException("Unexpected end of XML stream") else throw IOException("XML prolog or ROOT element not found on first $offset bytes") }
            if (offset == 0) return null
            input.reset(); val joined = BufferedReader(StringReader(prolog.substring(0, firstGt + 1))).readLines().joinToString(""); val matcher = ENCODING_PATTERN.matcher(joined); return if (matcher.find()) matcher.group(1).uppercase(Locale.ROOT).drop(1).dropLast(1) else null
        }
        private fun charset(name: String) = java.nio.charset.Charset.forName(name)
        @JvmStatic internal fun isAppXml(mime: String?): Boolean = mime != null && (mime == "application/xml" || mime == "application/xml-dtd" || mime == "application/xml-external-parsed-entity" || mime.startsWith("application/") && mime.endsWith("+xml"))
        @JvmStatic internal fun isTextXml(mime: String?): Boolean = mime != null && (mime == "text/xml" || mime == "text/xml-external-parsed-entity" || mime.startsWith("text/") && mime.endsWith("+xml"))
    }
}
